### PR TITLE
[Serving] Conversation template and OpenAI v1/chat/completions

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR supports the `v1/chat/completions` entrypoint of OpenAI API:
* protocol: `python/mlc_chat/protocol/openai_api_protocol.py`,
* entrypoint: `python/mlc_chat/serve/entrypoints/openai_entrypoints.py`

To support the chat functionality, this PR introduces the conversation template definition and register on Python side in `python/mlc_chat/serve/conversation.py`.
Compared with the current implementation of conversation template in C++, this version supports overriding the content of the system prompt without changing formatting texts (like the `[INST]` string in Llama2).

This PR adds a series of tests:
* temperature test for `v1/completions`,
* three different history message tests for `v1/chat/completions`,
* `max_tokens` test for `v1/chat/completions`,
* "wrong system prompt" position test for `v1/chat/completions`,
* the `openai` package test for both `v1/completions` and `v1/chat/completions`.

### CODE REVIEW

File: python/mlc_chat/serve/entrypoints/openai_entrypoints.py [hunk 1]
```diff
@@ -92,6 +103,10 @@ async def completion_stream_generator() -> AsyncGenerator[str, None]:
                 prompt, generation_cfg, request_id
             ):
                 num_completion_tokens += num_delta_tokens
+                if delta_text == "":
```
Comment (yzh119) [inline python/mlc_chat/serve/entrypoints/openai_entrypoints.py:106]: The delta only means insertion without delete, is that right? I wonder can we rename it to `appended_text` instead? (I know it's not introduced in this  PR, so we can adopt the renaming in later PRs.)

File: python/mlc_chat/serve/conversation.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+from typing import Dict, List, Optional, Tuple
+
+from pydantic import BaseModel, Field, field_validator
+
+# The system message placeholder in the system message prompt.
+SYSTEM_MESSAGE_PLACEHOLDER = "<<SYSTEM MESSAGE PLACEHOLDER>>"
+
+
+class Conversation(BaseModel):
```
Comment (yzh119) [inline python/mlc_chat/serve/conversation.py:11]: This is not related to the conversation template defined in our previous C++ codebase, is that correct? I wonder can we unify them?

File: python/mlc_chat/serve/config.py [hunk 3]
```diff
@@ -22,6 +22,8 @@ class GenerationConfig:
 
     max_new_tokens : int
         The maximum number of generated tokens.
+        It may be negative, which means the generation will not stop
```
Comment (yzh119) [inline python/mlc_chat/serve/config.py:25]: Is this design consistent with OpenAI's?

File: python/mlc_chat/protocol/protocol_utils.py [hunk 4]
```diff
@@ -25,13 +26,33 @@ def get_unsupported_fields(request: RequestProtocol) -> List[str]:
...
+    extra_stop_tokens: Optional[List[int]] = None,
+    extra_stop_str: Optional[List[str]] = None,
+) -> GenerationConfig:
     """Create the generation config in MLC LLM out from the input request protocol."""
-    if isinstance(request, OpenAIRequestProtocol):
-        return openai_api_get_generation_config(request)
-    raise RuntimeError("Cannot reach here")
+    kwargs: Dict[str, Any]
+    if isinstance(request, (OpenAICompletionRequest, OpenAIChatCompletionRequest)):
+        kwargs = openai_api_get_generation_config(request)
```
Comment (yzh119) [inline python/mlc_chat/protocol/protocol_utils.py:42]: Just curious, what's the difference between `openai_api_get_generation_config` and `get_generation_config`?

File: python/mlc_chat/protocol/openai_api_protocol.py [hunk 5]
```diff
@@ -98,40 +96,139 @@ class CompletionResponse(BaseModel):
...
+    model: str
+    system_fingerprint: str
+    object: Literal["chat.completion.chunk"] = "chat.completion.chunk"
+
+
+################################################
 
 
-def openai_api_get_unsupported_fields(request: OpenAIRequestProtocol) -> List[str]:
+def openai_api_get_unsupported_fields(
```
Comment (yzh119) [inline python/mlc_chat/protocol/openai_api_protocol.py:194]: When will this function be useful?

**Review by yzh119 — COMMENTED** [review]
Thank you for this effort, I left some comments/questions.

File: python/mlc_chat/serve/conversation.py (hunk 2 above)
Comment (tqchen) [reply python/mlc_chat/serve/conversation.py:11]: I think it would be good to have a common json schema(as being specified in here via pydantic), then embed all conv template in json. So we can only keep a c++ impl of the conv construction, all relies on a common spec.

The model here allows the API endpoint to also pass in conv template in future(this is a super set of openai api)

File: python/mlc_chat/serve/entrypoints/openai_entrypoints.py (hunk 1 above)
Comment (MasterJH5574) [reply python/mlc_chat/serve/entrypoints/openai_entrypoints.py:106]: “Delta” is used frequently I think. Our ChatModule uses the term “delta message,” and for example, OpenAI uses “delta” https://platform.openai.com/docs/api-reference/chat/streaming#chat/streaming-choices (you need to expand the entry “choices”. In this case, `appended_text` seems a bit unclear to me what “appended” means. I think we can go with the term “delta”.

File: python/mlc_chat/serve/config.py (hunk 3 above)
Comment (MasterJH5574) [reply python/mlc_chat/serve/config.py:25]: OpenAI supports `max_tokens = None` https://platform.openai.com/docs/api-reference/chat/create#chat-create-max_tokens, which means
> The total length of input tokens and generated tokens is limited by the model's context length.
 

In our case, “being negative means `None` in OpenAI” if we don’t use `std::optional`. Being negative is clear from my POV. I am not sure if using `std::optional` will incur additional runtime overhead.

File: python/mlc_chat/protocol/protocol_utils.py (hunk 4 above)
Comment (MasterJH5574) [reply python/mlc_chat/protocol/protocol_utils.py:42]: We may not only have OpenAI protocols. Say we have OpenAI protocols and our own protocols at the same time. Both protocols are defined in their respective files (like `python/mlc_chat/protocol/openai_api_protocol.py`), where their own conversion rules (like `openai_api_get_generation_config`) are implemented.

`get_generation_config` dispatches request to different protocol conversion rules.

File: python/mlc_chat/serve/config.py (hunk 3 above)
Comment (tqchen) [reply python/mlc_chat/serve/config.py:25]: I agree at the interface level we should just be consistent. aka support max_tokens = None and rename it to max_tokens

Comment (MasterJH5574) [reply python/mlc_chat/serve/config.py]: Updated accordingly. Now we use the name `max_tokens` and supports None in Python. On C++ side when we load the config from JSON, we check if the field value is `null`, and if so we use `-1` for `max_tokens` in C++.

**Review by tqchen — APPROVED** [review]
